File: Convert_an_Array_Into_a_2D_Array_With_Conditions.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var findMatrix = function(nums) {
    const res = [];
    const freq = {};

    for (const num of nums) {
        freq[num] = 1 + (freq[num] || 0);

        if (freq[num] > res.length) {
            res.push([]);
        }

        res[freq[num] - 1].push(num);
    }

    return res;    
};

File: Convert_an_Array_Into_a_2D_Array_With_Conditions.java
class Solution {
    public List<List<Integer>> findMatrix(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Map<Integer, Integer> freq = new HashMap<>();

        for (int num : nums) {
            freq.put(num, 1 + freq.getOrDefault(num, 0));

            if (freq.get(num) > res.size()) {
                res.add(new ArrayList<>());
            }

            res.get(freq.get(num) - 1).add(num);
        }

        return res;        
    }
}
